### ISSUE TITLE
MamaMsg leaks C-layer mamaMsg objects

### ISSUE THREAD

**Issue by dpauls** [OpenMAMA/OpenMAMA#258]
# ISSUE_TITLE
## Summary
*The CPP MamaMsg object can take ownership of the C-layer mamaMsg objects and then fail to destroy them, causing a memory leak.*

## Areas Affected
- [ ] MAMAC
- [X] MAMACPP
- [ ] MAMADOTNET
- [ ] MAMAJNI
- [ ] MAMDA
- [ ] MAMDACPP
- [ ] MAMDADOTNET
- [ ] MAMDAJNI
- [ ] Visual Studio
- [ ] SCons
- [X] Unit Tests
- [ ] Examples

## Details
Note: I started a discussion on this topic on the dev-list.  I receive messages in digest format, so my apologies if I am not addressing comments I may not have seen.

Perhaps the simplest way to illustrate the problem is with a unit test that triggers it.  I have created a simple unit test that does the following:
1. Creates a vector of 10 `MamaMsg` objects
2. Adds these `MamaMsg`s as a vector to another `MamaMsg`, `mainMsg`.
3. Calls `mainMsg->clear()`.
4. Deletes all of the above `MamaMsg` objects.
5. Repeats steps 1-4 1,000,000 times.

When I run the above unit test without any changes on Linux, the Linux oom-killer eventually kills the unit test.  Here is the source code for the unit test:
```
TEST_F (MamaMsgTestCPP, MemoryLeakTest)
{
    MamaMsg* msgs[10];
    MamaMsg* mainMsg = new MamaMsg();

    for (int i=0; i < 1000000; ++i) {
        for (int j=0; j < 10; ++j) {
            msgs[j] = new MamaMsg();
            msgs[j]->create();
        }

        mainMsg = new MamaMsg();
        mainMsg->create();
        mainMsg->addVectorMsg("foo", 1000, msgs, 10);
        mainMsg->clear();
        delete mainMsg;

        for (int j=0; j < 10; ++j) {
            delete msgs[j];
        }
    }
}
```

The following discusses the cause of the memory leak and a possible solution.

In a cpp `MamaMsg`, there are two members used to track the C-layer `mamaMsg` objects:
```
        mutable mamaMsg*       mCvectorMsg;
        mutable size_t         mCvectorMsgAllocSize;
```

By looking at the `MamaMsg` code, it appears their purpose is to take ownership of the C-layer objects contained in `MamaMsg[]` objects passed to `MamaMsg::addVectorMsg()`.  Then when the `MamaMsg` that contains the vector of messages is destroyed, the C-layer `mamaMsg` objects are destroyed at that time.

However, looking at the `MamaMsg` code itself, I see no reason for it to do this.  The origins of this code go back to commit 571dfea3, which is the original implementation from 2012.  I don't see any comments or clues as to why this was done.

My best theory is that it is to allow the caller of `MamaMsg::addVectorMsg` to delete the `MamaMsg[]`, and yet *allow lower layers to retain references to the C-layer `mamaMsg` objects*.  I’m skeptical whether this is required.  By looking at the C API, there is no such protection:
* An application could call `mamaMsg_addVectorMsg()`
* The application could then `mamaMsg_destroy()` each `mamaMsg` in the value array passed in.

So for consistency of semantics between the C and CPP interfaces, additional tracking should not be added in the CPP layer.

In our scenario with the Solace payload bridge, I’ve tried a patch that removes `mCvectorMsg` and `mCvectorMsgAllocSize`.  I don’t observe any problems.  Does this sound like an appropriate fix?

If not, there are numerous issues that need to be investigated and fixed.  I’ll outline a few of them, to the degree that I understand them right now:
1.	`MamaMsg::clear()` destroys the `mCvectorMsg` array without destroying the `mamaMsg` objects contained within (it only calls `mamaMsg_clear()` on them).  Unfortunately, simply adding `mamaMsg_destroy()` doesn’t solve the problem.  When we did this, a valgrind test lit up several invalid memory access and invalid free errors.  More investigation would be required here.
2.	If `MamaMsg::addVectorMsg` is called multiple times, the contents of `mCvectorMsg` are overwritten without any `mamaMsg_destroy()` on the objects before being overwritten.  A naïve attempt to fix this would be to simply destroy them before overwriting them.  But I don’t think we’d have any guarantee that the new messages are different than the old ones.  It feels like something like a refCount is really required (and if that’s the case, I think the refCount should be where the reference might actually be required).  You _could_ linearly search to see if the added values match any existing values and free only the ones that aren’t present in the new vector, but that is getting to be an O(n^2) algorithm, and it _still_ doesn’t solve all of the issues, see #3 below.
3.	`MamaMsg` only stores a single `CvectorMsg`.  What if a vector of messages is added to more than one fieldId in the message?  In order to really address this, I think you’d need to maintain a map of mamaMsg pointers to fieldId’s that reference those mamaMsgs.  Then only destroy the mamaMsg when there are no fieldId’s left referencing it.
4.	What if an application follows a pattern demonstrated in the unit test above?  How should this be handled?

Due to the numerous issues above with transferring ownership of `mamaMsg` objects as part of a `MamaMsg::addVectorMsg()`, and because I don’t see a compelling reason to do it in the first place, I believe the correct fix is to remove these semantics completely.

Is anyone aware of a reason this should not be changed this way?

If this sounds like a reasonable approach, I will submit a PR with the change and associated unit test shortly.

**Comment by dpauls** [comment]
I have implemented and tested a set of changes in https://github.com/dpauls/OpenMAMA/tree/issue-258.  I can open a PR if this looks like the right approach.

Unit tests, including the new unit test pass with these changes.

Before submitting the PR though, I vaguely recall not doing something correctly last time related to signing the PR.  I'm still trying to figure out how that is to be done.

**Comment by fquinner** [comment]
Since this has been there for a long time and it's a little scary, I'd say this could wait until 6.2.1.

Go ahead and raise the PR if you already have the work done - we can always adjust the approach later if there is some unknown reason for this approach that surfaces. The signed off by bit is done in the commit note: https://github.com/OpenMAMA/OpenMAMA/wiki/Commit-Guidelines#signed-off-by

**Comment by fquinner** [comment]
Merged in #259 - thanks @dpauls
